### PR TITLE
Feature/cosmosdb

### PR DESCRIPTION
**Pull request by abouroubi**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #27 

## What is the new behavior?

Add basic functions for Cosmos DB

## Does this PR introduce a breaking change?
```
[ ] Yes
[X ] No
```

## Other information

### CODE REVIEW

**Comment by santoshyadav198613** [comment]
Thanks for PR @abouroubi  I will start reviewing it.

File: lib/cosmos-db/cosmos-db.repository.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
+// import { Inject, Injectable, Logger } from '@nestjs/common';
+// import { AZURE_COSMOS_DB_CONNECTION_NAME } from './cosmos-db.constants';
+// import {
+//   //AzureTableContinuationToken,
+//   AzureCosmosDbQuery,
+//   AzureCosmosDbResponse,
+//   AzureCosmosDbResultList,
+//   Repository,
+// } from './cosmos-db.interface';
+// import { AzureEntityMapper } from './cosmos-db.mapper';
```
Comment (santoshyadav198613) [inline lib/cosmos-db/cosmos-db.repository.ts]: We can remove this file if not needed.

File: lib/cosmos-db/cosmos-db.constants.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export const AZURE_COSMOS_DB_MODULE_OPTIONS = 'AZURE_COSMOS_DB_MODULE_OPTIONS';
```
Comment (santoshyadav198613) [inline lib/cosmos-db/cosmos-db.constants.ts]: COSMOS_DB_MODULE_OPTIONS = 'COSMOS_DB_MODULE_OPTIONS';

File: lib/cosmos-db/cosmos-db.constants.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+export const AZURE_COSMOS_DB_CONNECTION_NAME = 'AZURE_COSMOS_DB_CONNECTION_NAME';
```
Comment (santoshyadav198613) [inline lib/cosmos-db/cosmos-db.constants.ts]: COSMOS_DB_CONNECTION_NAME = 'COSMOS_DB_CONNECTION_NAME';

File: lib/cosmos-db/cosmos-db.constants.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+export const AZURE_COSMOS_DB_NAME = 'AZURE_COSMOS_DB_NAME';
```
Comment (santoshyadav198613) [inline lib/cosmos-db/cosmos-db.constants.ts]: COSMOS_DB_NAME = 'COSMOS_DB_NAME';

File: lib/cosmos-db/cosmos-db.constants.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+export const DEFAULT_DB_CONNECTION = 'AZURE_COSMOS_DB_DEFAULT_CONNECTION';
```
Comment (santoshyadav198613) [inline lib/cosmos-db/cosmos-db.constants.ts]: DEFAULT_DB_CONNECTION = 'COSMOS_DB_DEFAULT_CONNECTION';

File: lib/cosmos-db/cosmos-db.decorators.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
+import { Inject } from '@nestjs/common';
+import { getConnectionToken, getModelToken } from './cosmos-db.utils';
+
+export const AZURE_COMSOS_DB_ENTITY = 'azure-cosmos-db:entity';
```
Comment (santoshyadav198613) [inline lib/cosmos-db/cosmos-db.decorators.ts]: COMSOS_DB_ENTITY = 'cosmos-db:entity';

**Comment by santoshyadav198613** [comment]
Awesome work @abouroubi . few comments especially on naming convention as we are not exposing them as public API.

**Comment by santoshyadav198613** [comment]
Also for partition key refer this PR https://github.com/nestjs/azure-database/pull/32/files we may get something like this in future for this API as well, check if we can get some implementation around same.

**Comment by abouroubi** [comment]
> Also for partition key refer this PR https://github.com/nestjs/azure-database/pull/32/files we may get something like this in future for this API as well, check if we can get some implementation around the same.

Not sure about this one, if we add a composite key, we must add it to all the create queries, which is not possible because we expose the CosmosDb container directly to the user so we are not able to to preprocess the create action to add the Partition Key to each query.

File: README.md [hunk 7]
```diff
@@ -181,7 +187,8 @@ import { Contact } from './contact.entity';
 export class ContactService {
   constructor(
     @InjectRepository(Contact)
-    private readonly contactRepository: Repository<Contact>) {}
+    private readonly contactRepository: Repository<Contact>,
```
Comment (santoshyadav198613) [inline README.md]: private readonly contactRepository: Repository<Contact>

File: README.md [hunk 8]
```diff
@@ -279,6 +286,213 @@ The `AzureTableStorageRepository` provides a couple of public APIs and Interface
...
+  date: Date;
+  location: Point;
+}
+```
+
+2. Create a file called `event.entity.ts` and describe the entity model using the provided decorators:
+
+- `@CosmosPartitionKey(value: string)`: Represents the `PartitionKey` of the entity (**required**).
+
+- `@CosmosDateTime(value?: string)`: For time of day.
```
Comment (santoshyadav198613) [inline README.md]: for DateTime values.

**Comment by santoshyadav198613** [comment]
Hi @abouroubi ,
Thanks for making all the changes, just one last thing and we are done, as this will be a new feature, lets do 2 things as well:
1. Add one example in root create an example folder and add one nestjs project with cosmos DB implementation.
2. Also a blog post about this feature.

Let me know if you need any help in any of these 2.

**Comment by abouroubi** [comment]
Hi @santoshyadav198613,

Yeah I can do that, I can write a blog post, but I don't have a blog to publish it

**Comment by santoshyadav198613** [comment]
That's not a problem, we can use official blog post of NestJS, can you write it using medium as of now, get the draft ready.

**Comment by abouroubi** [comment]
I added the sample, I'll write the blog post, in the upcoming days

**Comment by abouroubi** [comment]
Can you publish a preversion, so we can test sample app with the correct imports ?
I've tested it locally, but it's always better to test with the real npm package

**Comment by santoshyadav198613** [comment]
Cool @manekinekko can we get the beta version released code looks good to me.

**Comment by santoshyadav198613** [comment]
@abouroubi , are you on Twitter,  can you share your handle if possible?

**Comment by abouroubi** [comment]
It's @abdoubouroubi but I'm not really present there

**Comment by abouroubi** [comment]
@santoshyadav198613 @manekinekko any news ?

**Comment by santoshyadav198613** [comment]
Hey @abouroubi, really sorry busy will get it merged soon.

**Comment by mnokeefe** [comment]
Any news on this?

**Comment by abouroubi** [comment]
I don't have any news either

**Comment by manekinekko** [comment]
Hi guys, sorry for the delay. I was on paternity leave for the last 2 months.
I am gonna review the PR this week and will make sure to cut a release in the coming weeks.

**Comment by DenzoNL** [comment]
Code looks good to me too. @manekinekko Would you be willing to push out a beta release today? I'll get my hands dirty on testing this.

**Comment by manekinekko** [comment]
@DenzoNL I will merge this into master but can't cut a release on NPM yet. You will be able to try it out from master until a release is available on NPM.

Will that be ok?

**Comment by DenzoNL** [comment]
@manekinekko That would be great for now :)

**Pull request merged**